### PR TITLE
Use git/hg to determine correct commit for snapshot

### PR DESCRIPTION
**Pull request by tarsius**
This is the second in a series of pull-requests leading up to the implementation of a new version number scheme.  This also includes the commits from the first pull-request in the series; discuss those at #66.

The changes in these commits are not *required* to implement the new version number scheme.  However, changing *how we determine* the commit to be packaged, and what version string we use to identify that package, are related.

Both of these changes (can) lead to version strings decreasing, so it seems like a good idea to merge them at the same time.  I therefore that we do not merge this into `master` before we merge the new version number scheme, and that we not even use it on Melpa itself.

But while related, these are separate changes, which should be discussed separately.

This commit fixes the issue raised in #61: in some edge-cases we do not identify the correct commit to build a package from.  Since these are edge-cases, we could ignore them, which would be appealing because fixing them results in a breaking change, but I think that now would be a good time to address them, along with the other big breaking change that is switching to a new version number scheme.

The change, including the trade offs are described in detail in the commit message:

```
Use git/hg to determine correct commit for snapshot

Before this commit we determined the files that match a package's
files-spec using an elisp implementation and then we used git/hg to
determine the latest commit in which one of these files was modified.
This had two issues:

1. It was necessary to check out the latest upstream commit as part
   of the commit selection step.

   This led to an unfortunate lack of separation of concerns, most
   of which were addressed in commits leading up to this one.

   The remaining lack of separation could be addressed by using a
   new elisp implementation, which operated on a git tree instead of
   on checked out files, but doing that would not address the second
   issue below.  Addressing that, would mean throwing out this new
   implementation in the very next commit, which is why this commit
   tackles these two issues at once.

2. The wrong commit was selected in at least the following edge-case:

   A first commit A touches a matching file.  A second commit B
   removes another matching file.  A third and final (HEAD) commit C
   doesn't modify any matching files, so no snapshot package is build
   for that commit.

   Never-the-less C is used to determine what the matching files are
   and because the file that was removed in B, no longer exists in C
   we concluded that A was the last commit in which something relevant
   happened.  The removal of a matching file in B warrants a new
   snapshot based on that commit, but because HEAD is at C, we used
   to built a snapshot for A instead.

The second issue could be addressed in elisp alone, i.e., we could
walk all commits without relying on git and hg, but that would be
very inefficient and likely error-prone.

This commit separates concerns between elisp and git/hg differently.
`package-build--spec-globs' converts the files spec into arguments
understood by git or hg.  Git or hg then takes care of finding the
last commit that touched any matching file.  This includes removals
of matching files.

Unfortunately both git and hg process all include rules before all
exclude rules, i.e., they disregard the order of the arguments.
Previously it was possible to override the default exclude rules in
the recipe-specific files spec, but we have to give up on that.  In
practice this change affects very few packages, five to be precise,
but it *is* a change in behavior, and it is forced upon us by this
limitation in git and hg.

Determining the latest commit that touches any of the matching files
and actually listing the matching files in that commit are separate
tasks, which were always performed separately.  It would have been
nice to also rely on git resp. hg to do the latter, but unfortunately
that is not possible.

While git and hg match the same files as the elisp implementation
(modulo the exclude complication), they do not produce the same
output.  When a directory matches a glob, then they list all files
within that directory instead of the directory itself.  That makes
sense from their perspective, but is not what we need.  So we still
need `package-build--expand-files-spec' (and have to adjust it to
process includes before excludes, to match the git/hg behavior).
```

Speaking of `hg`, there is one more change I would like to discuss.  I have already added *optional* support for using `git-remote-hg` instead of `hg` directly to clone and pull Mercurial repositories. (`git-remote-hg` is an implementation of Git's "remote-helper" protocol, which allows interacting with foreign VCS using `git`.)

I would like to suggest that we use that for Melpa itself but keep support for using `hg` directly for functionality that *already exists*.  That would mean that I would not have to implement parts of the code required for by the new version number scheme twice.  We could still implement it for `hg` *eventually*, but not requiring this up front, would allow us to move faster.

I have used `git-remote-hg` on the Emacsmirror for years.  It works very reliably.  The only drawbacks are that it will have to be installed on the server and that the commit hashes we record are Git's hashes of the imported commits, not the hashes used by Mercurial itself.  The recorded hashes are mostly intended for internal purposes, so I would say this is not a big deal.  Also there are only 14 packages that use Mercurial.

### CODE REVIEW

**Comment by tarsius** [comment]
This fails for `edts`.

Firstly, congratulations for picking that package as one to use for tests. :smiley: 

Secondly, this is no surprise. This is one of the four recipes I have already identified as requiring changes: https://github.com/melpa/melpa/compare/fix-snapshots.

By the way, while @houseofsuns and I have already tested earlier versions of this change to see if any packages broke in unexpected ways, there have been enough changes to the implementation to make it necessary to do that again.

**Comment by tarsius** [comment]
> This is one of the four recipes I have already identified as requiring changes: https://github.com/melpa/melpa/compare/fix-snapshots.

By the way, I would like to merge that now, even though it isn't *necessary* until/if this is merged. Doing so would make testing easier, as I wouldn't have to deal with false-positives anymore.

File: package-build.el
Comment (milkypostman) [inline package-build.el:231]: Any idea what timezone `hg` will output? Do we need to convert to UTC?

A comment expressing that we ignore the timezone and assume UTC may be worth it.

**Comment by milkypostman** [comment]
> Unfortunately both git and hg process all include rules before all
exclude rules, i.e., they disregard the order of the arguments.
Previously it was possible to override the default exclude rules in
the recipe-specific files spec, but we have to give up on that.

This seems fine.

I'm reviewing the change now.

File: package-build.el
Comment (milkypostman) [inline package-build.el:723]: is this actually different now?

File: package-recipe.el
Comment (milkypostman) [inline package-recipe.el:196]: Lets comment above this "Ignore `:defaults` which is valid".

File: package-recipe.el
Comment (milkypostman) [inline package-recipe.el:200]: Commenting this section would help as well. Maybe just saying that we're verifying all the globs are strings?

File: package-recipe.el
Comment (milkypostman) [inline package-recipe.el:50]: This change seems like we're mixing abstractions. `package-recipe` to me is a representation of the actual recipe. But then we process the recipe and get something else, which is more like a `package-snapshot` (not a well thought out name) that includes the "version" and "time" or something. 

Is this something we should work towards? I.e., it would make sense that in the future we have a function that takes a recipe as input and returns a snapshot. Both of which are not mutable.

File: package-build.el
Comment (milkypostman) [inline package-build.el:807]: This function deserves a small comment. E.g.,

> Produce arguments from RCP for including and excluding files for the relevant VCS.

File: package-build.el
Comment (milkypostman) [inline package-build.el]: It wouldn't hurt to add a comment for each case,

```
;; top level include globs
```

```
;; excluded globs
```

```
;; globs for a subdirectory
```

File: package-build.el
Comment (milkypostman) [inline package-build.el:767]: `Return a list of all files matching SPEC in SUBDIR`?

File: package-build.el
Comment (milkypostman) [inline package-build.el:753]: `Return a list of all files matching SPEC in default-directory.`?

File: package-build.el
Comment (milkypostman) [inline package-build.el:625]: Function comment `Get the package description for RPC potentially searching FILES.`?

It may be worth adding more high-level description here in the function comment as this is an important part of code that has been recently discussed a lot.

File: package-build.el
Comment (milkypostman) [inline package-build.el:311]: Please comment this function.

```
Runs COMMAND in the default directory.

We use this to wrap commands is proper environment settings and with a timeout so that no command can block the build process.
```

**Review by milkypostman — COMMENTED** [review]
As time has gone on I'm getting more dependent on commenting, even when it's a bit trivial, so that I can quickly skim code and understand it. So that is the majority of my comments, besides working through the change.

File: package-build.el
Comment (tarsius) [reply package-build.el:231]: I've added this comment:

```
                     ;; The "date" keyword uses UTC. The "hgdate" filter
                     ;; returns two integers separated by a space; the
                     ;; unix timestamp and the time zoneoffset.  We use
                     ;; "hgdate" because that makes it easier to discard
                     ;; the time zone offset, which doesn't interest us.
```

The timezone offset is for the timezone in which to commit was created. I am not adding this information to the comment because we don't care about the timezone.

File: package-build.el
Comment (tarsius) [reply package-build.el:723]: Yes, previously it was possible to use nested SPECs. That feature was used by about half a dozen recipes and made them less readable.

File: package-recipe.el
Comment (tarsius) [reply package-recipe.el:194]: It is not invalid. I have added this comment:

```
        ;; `:defaults' is only allowed as the first element.  If we find
        ;; it in that position, just skip over it, keeping the subsequent
        ;; validation simpler.
```

(This seems a bit much too me.)

Comment (tarsius) [reply package-recipe.el:196]: I've added these comments:

```
    (when-let ((spec (plist-get plist :files)))
        ;; `:defaults' is only allowed as the first element.
        ;; If we find it in that position, skip over it.
        (when (eq (car spec) :defaults)
          (setq spec (cdr spec)))
        ;; All other elements have to be strings or lists of strings.
        ;; A list whose first element is `:exclude' is also valid.
        (dolist (entry spec)
```

File: package-recipe.el
Comment (tarsius) [reply package-recipe.el:200]: Done.

File: package-recipe.el
Comment (tarsius) [reply package-recipe.el:50]: > This change seems like we're mixing abstractions. `package-recipe` to me is a representation of the actual recipe.

Change your mental modal to "build instructions". We start out with the recipe file, which explains how any version of that package has to be build. We then determine which version we want to build. This becomes part of the instructions the code that actually does the building has to follow.

Previously we passed this additional information along as separate arguments. As noted in the commit message this complicated function signatures, and also led to some information not being passed along and instead being recalculated later on. Sticking to separate arguments also means that any change to what information is being passed along, is a breaking change. In the past such changes did affect public functions, going forward it shouldn't happen anymore.

> Is this something we should work towards? I.e., it would make sense that in the future we have a function that takes a recipe as input and returns a snapshot. Both of which are not mutable.

This feels like an unnecessary complication. There isn't just one recipe class but one for each type of upstream repository/forge that we support. We would have to add one "snapshot" class for each "recipe" class. Adding two slots that do not appear in recipe files to the base recipe class, seems like a much cleaner way of implementing this and I don't agree that doing that means that we are mixing abstractions.

**Comment by tarsius** [comment]
I've added the requested docstrings and comments (d0fe7299b36e029a52c999b0acad6895d0327598) but haven't spread those changes across the appropriate commits yet.

**Comment by milkypostman** [comment]
Some checks failed. But this looks good to me.

**Review by milkypostman — APPROVED** [review]

**Comment by tarsius** [comment]
@melpa/maintainers It's-a merge!

The last commit ensures that a snapshot is never older than the latest release. If the last commit that touched a relevant file is reachable from the latest release, then the snapshot is created from the released commit (but still using a timestamp). This leads to more than a hundred packages being updated.

I tested *a lot*, which did uncover a few bugs in the new code but also in the previous code and in recipes that didn't work right given the old or new code. (Such as including all of `.git` in the package. :scream:)

If we get any reports about files that are missing from packages, or files that shouldn't be include, etc., then please ping me.

On the whole, about twenty recipes had to be adjusted, to account for changes in how `:files` is used. Packages whose name end with `-test` and libraries (non-tests) whose name end with `-test.el` are problematic now, `:defaults` cannot be used in those cases. I've updated Melpa's readme accordingly.

**Comment by milkypostman** [comment]
Thanks @tarsius for the message. 🤞 

I'm glad we got this in. 

So if there is a commit to a file that would lead to a decrease in timestamp version are you saying we reuse the most recent timestamp version?

**Comment by tarsius** [comment]
> So if there is a commit to a file that would lead to a decrease in timestamp version are you saying we reuse the most recent timestamp version?

There can (no longer) be a "decrease in timestamp version". Previously it was possible that the removal of a relevant file could cause an older commit to be considered the most recent commit worth creating a snapshot for. Now that we use git/hg to determine the last commit that touched a relevant file (including removal of such a file), this cannot happen anymore.

What I mean is this

```
dddddd Fix typo in README
cccccc master 1.0.0 Release Foo 1.0.0  
bbbbbb Update README
aaaaaa Edit foo.el
```

Previously we would have created a snapshot from `aaaaaa` because that is the last commit that touches `foo.el`. Now we create a snapshot from `cccccc` to ensure that the snapshot is not "older" than the latest release. But in the following case we don't do that.

```
dddddd * master Fix typo in README
cccccc |  * release 1.0.0 Release Foo 1.0.0
       |/
bbbbbb * Update README
aaaaab * Edit foo.el
```

**Pull request merged**
